enterprise/server/backends/userdb/userdb_test.go

```diff
@@ -387,32 +448,93 @@ func TestUpdateGroupUsers_Role(t *testing.T) {
-	us1Group := u.Groups[0].Group
+	us1Group := getSelfOwnedGroup(t, ctx1, env).Group
 
-	// Make sure we can update their role within the group
-	err = udb.UpdateGroupUsers(ctx, us1Group.GroupID, []*grpb.UpdateGroupUsersRequest_Update{{
+	err := udb.UpdateGroupUsers(ctx1, us1Group.GroupID, []*grpb.UpdateGroupUsersRequest_Update{{
 		UserId: &uidpb.UserId{Id: "US1"},
 		Role:   grpb.Group_DEVELOPER_ROLE,
 	}})
-	require.NoError(t, err)
+	require.NoError(t, err, "US1 should be able to update their own group role")
 
 	groupUsers, err := udb.GetGroupUsers(ctx1, us1Group.GroupID, []grp.GroupMembershipStatus{grp.GroupMembershipStatus_MEMBER})
 	require.NoError(t, err)
 	us1 := findGroupUser(t, "US1", groupUsers)
-	require.Equal(t, grpb.Group_DEVELOPER_ROLE, us1.Role, "user role should be DEVELOPER")
+	require.Equal(t, grpb.Group_DEVELOPER_ROLE, us1.Role, "US1 role should be DEVELOPER")
+
+	createUser(t, ctx, env, "US2", "org2.io")
+	ctx2 := authUserCtx(ctx, env, t, "US2")
+
+	_, err = udb.GetGroupUsers(ctx2, us1Group.GroupID, nil /*=status*/)
+	require.True(
+		t, status.IsPermissionDeniedError(err),
+		"expected PermissionDeniedError if US2 tries to list US1's group users; got: %T",
+		err)
+
+	err = udb.UpdateGroupUsers(ctx2, us1Group.GroupID, []*grpb.UpdateGroupUsersRequest_Update{{
+		UserId: &uidpb.UserId{Id: "US1"},
+		Role:   grpb.Group_ADMIN_ROLE,
+	}})
+	require.True(
+		t, status.IsPermissionDeniedError(err),
+		"expected PermissionDeniedError if US2 tries to update US1's group role; got: %T",
+		err)
+}
+
+func TestUpdateAPIKey(t *testing.T) {
+	ctx := context.Background()
+	env := newTestEnv(t)
+	flags.Set(t, "app.create_group_per_user", true)
+	flags.Set(t, "app.no_default_user_group", true)
+	udb := env.GetUserDB()
+
+	createUser(t, ctx, env, "US1", "org1.io")
+	ctx1 := authUserCtx(ctx, env, t, "US1")
+	gr1 := getSelfOwnedGroup(t, ctx1, env)
+
+	k1 := getSingleAPIKey(t, ctx1, env, gr1.Group.GroupID)
+	k1.Label = "US1-Updated-Label"
+	err := udb.UpdateAPIKey(ctx1, k1)
+
+	require.NoError(t, err, "US1 should be able to update their own API key")
+
+	createUser(t, ctx, env, "US2", "org2.io")
+	ctx2 := authUserCtx(ctx, env, t, "US2")
+
+	k1.Label = "US2-Updated-Label"
+	err = udb.UpdateAPIKey(ctx2, k1)
+
+	require.True(
+		t, status.IsPermissionDeniedError(err),
+		"expected PermissionDeniedError if US2 tries to update US1's API key; got: %T",
+		err)
+}
+
+func TestDeleteAPIKey(t *testing.T) {
+	ctx := context.Background()
+	env := newTestEnv(t)
+	flags.Set(t, "app.create_group_per_user", true)
+	flags.Set(t, "app.no_default_user_group", true)
+	udb := env.GetUserDB()
+
+	createUser(t, ctx, env, "US1", "org1.io")
+	ctx1 := authUserCtx(ctx, env, t, "US1")
+	gr1 := getSelfOwnedGroup(t, ctx1, env)
+
+	k1 := getSingleAPIKey(t, ctx1, env, gr1.Group.GroupID)
+	k1.Label = "US1-Updated-Label"
+	err := udb.UpdateAPIKey(ctx1, k1)
+
+	require.NoError(t, err, "US1 should be able to update their own API key")
```

Comment:
Should this test to see if a user can delete their own API key? If not, this section should probably just be removed, since it is already tested above in the relatedly-named function.

Reply:
Thanks for calling this out! Turns out this PR actually introduced a deadlock in `DeleteUser()` caused by trying to start a new transaction within the existing transaction. Updated this test and fixed the bug.